x/ibc/03-connection/keeper/keeper.go

```diff
@@ -0,0 +1,183 @@
+	return Keeper{
+		storeKey:     key,
+		cdc:          cdc,
+		codespace:    sdk.CodespaceType(fmt.Sprintf("%s/%s", codespace, types.DefaultCodespace)), // "ibc/connections",
+		prefix:       []byte(types.SubModuleName + "/"),                                          // "connections/"
+		clientKeeper: ck,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s/%s", ibctypes.ModuleName, types.SubModuleName))
+}
+
+// GetConnection returns a connection with a particular identifier
+func (k Keeper) GetConnection(ctx sdk.Context, connectionID string) (types.ConnectionEnd, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := store.Get(types.KeyConnection(connectionID))
+	if bz == nil {
+		return types.ConnectionEnd{}, false
+	}
+
+	var connection types.ConnectionEnd
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &connection)
+	return connection, true
+}
+
+// SetConnection sets a connection to the store
+func (k Keeper) SetConnection(ctx sdk.Context, connectionID string, connection types.ConnectionEnd) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(connection)
+	store.Set(types.KeyConnection(connectionID), bz)
+}
+
+// GetClientConnectionPaths returns all the connection paths stored under a
+// particular client
+func (k Keeper) GetClientConnectionPaths(ctx sdk.Context, clientID string) ([]string, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := store.Get(types.KeyClientConnections(clientID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var paths []string
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &paths)
+	return paths, true
+}
+
+// SetClientConnectionPaths sets the connections paths for client
+func (k Keeper) SetClientConnectionPaths(ctx sdk.Context, clientID string, paths []string) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), k.prefix)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(paths)
+	store.Set(types.KeyClientConnections(clientID), bz)
+}
+
+// addConnectionToClient is used to add a connection identifier to the set of
+// connections associated with a client.
+//
+// CONTRACT: client must already exist
+func (k Keeper) addConnectionToClient(ctx sdk.Context, clientID, connectionID string) sdk.Error {
+	conns, found := k.GetClientConnectionPaths(ctx, clientID)
+	if !found {
+		return types.ErrClientConnectionPathsNotFound(k.codespace)
+	}
+
+	conns = append(conns, connectionID)
+	k.SetClientConnectionPaths(ctx, clientID, conns)
+	return nil
+}
+
+// removeConnectionFromClient is used to remove a connection identifier from the
+// set of connections associated with a client.
+//
+// CONTRACT: client must already exist
+func (k Keeper) removeConnectionFromClient(ctx sdk.Context, clientID, connectionID string) sdk.Error {
```

Comment:
Should this check for client existence and return an error if no?

Reply:
The spec ensures this contract, but it's up to the SDK team